homeassistant/components/smartthings/binary_sensor.py

```diff
@@ -53,9 +53,17 @@ async def async_setup_entry(
     broker = hass.data[DOMAIN][DATA_BROKERS][config_entry.entry_id]
     sensors = []
     for device in broker.devices.values():
-        for capability in broker.get_assigned(device.device_id, "binary_sensor"):
-            attrib = CAPABILITY_TO_ATTRIB[capability]
-            sensors.append(SmartThingsBinarySensor(device, attrib))
+        device_capabilities_for_binary_sensor = broker.get_assigned(
+            device.device_id, "binary_sensor"
+        )
+        for component in device.components:
+            if component in device.status.disabled_components:
+                continue
+            for capability in device.components[component]:
+                if capability not in device_capabilities_for_binary_sensor:
+                    continue
+                attrib = CAPABILITY_TO_ATTRIB[capability]
```

Comment:
All of this logic, including the check for `disabledComponents` and `disabledCapabilities` should be moved inside `broker.get_assigned`. This method, and the supporting logic should be modified to operate on a component.

Reply:
That's actually how I started, but it turned out to be a large refactor and I moved to this method instead. But I'll give it another go.